concurrency/ApplicationEntitiesSearcher.java

```diff
@@ -0,0 +1,245 @@
+
+        return SearchApplicationsMatch.builder()
+                .withPermissionGroup(PermissionGroup.builder()
+                        .withPermissionGroupName(entity.getOwnedByPermissionGroupName())
+                        .withPermissionGroupType(entity.getOwnedByPermissionGroupType().toUpperCase())
+                        .build())
+                .withApplicationId(entity.getApplicationID())
+                .withApplicationName(entity.getApplicationName())
+                .withApplicationVersion(application.getApplicationVersion())
+                .withApplicationStatus(entity.getStatus())
+                .build();
+    }
+
+    private SearchServicesMatch mapServices(final ApplicationEntity entity,
+                                            final ConcurrentHashMap<String, Application> applications) {
+        final Application application =
+                getApplication(entity.getApplicationID(), applications);
+
+        return SearchServicesMatch.builder()
+                .withApplicationId(entity.getApplicationID())
+                .withApplicationVersion(application.getApplicationVersion())
+                .withApplicationName(entity.getApplicationName())
+                .withServiceStatus(entity.getStatus())
+                .withRelationshipType(DBAccessor.
+                        convertApplicationEntityTypeToRelationshipType(entity.getType()))
+                .withServiceName(entity.getValue())
+                .build();
+    }
+
+    private SearchClientsMatch mapClients(final ApplicationEntity entity,
+                                          final ConcurrentHashMap<String, Application> applications) {
+        final Application application =
+                getApplication(entity.getApplicationID(), applications);
+
+        return SearchClientsMatch.builder()
+                .withApplicationId(entity.getApplicationID())
+                .withApplicationVersion(application.getApplicationVersion())
+                .withApplicationName(entity.getApplicationName())
+                .withClientStatus(entity.getStatus())
+                .withRelationshipType(DBAccessor.convertApplicationEntityTypeToRelationshipType(entity.getType()))
+                .withClientName(entity.getValue())
+                .build();
+    }
+
+    /**
+     * Gets an application that was already retrieved or from the database.
+     * @param applicationID
+     * @param applications
+     * @return
+     */
+    private Application getApplication(final String applicationID, final ConcurrentHashMap<String, Application> applications) {
+        if(!applications.containsKey(applicationID)) {
+            final Application application = this.applicationByApplicationIdGetter
+                    .apply(LatestVersionRequest.<String>builder()
+                            .request(applicationID)
+                            .build());
+            applications.put(applicationID,application);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `containsKey()` and `put()` may not be thread-safe at lines: **235 and 240**. Two threads can perform this **same check at the same time** and one thread can **overwrite** the value written by the other thread.

**Fix**  
 Consider replacing `put()` with `putIfAbsent()` to help prevent accidental overwriting. `putIfAbsent()` puts the value only if the `ConcurrentHashMap` does not contain the key and therefore avoids overwriting the value written there by the other thread&#39;s `putIfAbsent()`.

**More info**  
 `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.

 [View an example on GitHub](https://github.com/apache/hadoop/blob/f16cf877e565084c66bc63605659b157c4394dc8/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/startupprogress/StartupProgress.java#L272-L273) (external link).